presto-spi/src/main/java/com/facebook/presto/spi/function/SqlInvokedRegularFunction.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.function;
+
+import com.facebook.presto.spi.relation.FullyQualifiedName;
+import com.facebook.presto.spi.type.TypeSignature;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.facebook.presto.spi.function.FunctionKind.SCALAR;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.collectingAndThen;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toList;
+
+public class SqlInvokedRegularFunction
+        implements SqlFunction, BoundFunction
```

Comment:
This looks fishy. Conceptually a `BoundFunction` should be a `SqlFunction`.

Reply:
Which classes with suffix `Function` are not `SqlFunction`?

Reply:
Maybe it makes more sense to rename `SqlFunction` to `GenericSqlFunction` and `BoundFunction` should be `SqlFunction` and `specialize` should return a `SqlFunction`. That way you unifies both builtin and sql functions and both are first class citizen and the definition is clear.